[mono][swift-interop] Support for Swift frozen struct lowering in mini-JIT/AOT/interpreter

matouskozak opened the pull request:
## Description
This PR adds support for lowering of Swift `@frozen` structs in P/Invoke arguments. In short, in Swift calls, `@frozen` structs can be passed as a sequence of up to 4 primitives using registers or stack (see [Swift Calling Convention](https://github.com/apple/swift/blob/main/docs/ABI/CallingConvention.rst#general-expansion-algorithm)). The general algorithm was implemented in https://github.com/dotnet/runtime/pull/99439 and this PR connects it to the Mono codegen backends. This PR doesn't handle returns of Swift `@frozen` structs and reverse P/Invoke scenarios (tracked at https://github.com/dotnet/runtime/issues/102077).

## Changes
The connection to codegen is done at Mono IR level. First, in `method-to-ir.c` for connection to JIT and AOT backends. Second, in `transform.c` for interpreter. Both connection are similar. 

Before doing call to native Swift code, we explore the signature for any struct types. If struct encountered (not `SwiftSelf` or `SwiftError` which are handled individually), we execute the lowering algorithm. Based on the result we have 2 outcomes:
1. **Structure can be lowered into a sequence of up to 4 primitives:**
The structs argument in signature is replaced by the lowered sequence. The sequence elements are taken from the original structs at previously calculated offsets (struct base address + element offset). For detail how are the offsets calculated see `mono_marshal_get_swift_physical_lowering`.

2. **Structure must be passed by reference:**
The structs argument in signature is replaced by a byref type that represents address of the structs.

Non-struct types are preserved. The modified signature is used just for the call to native Swift code (for non-P/Invoke calls original signature is used).

---

Lastly, refactoring and bug fixes for marshaling inside `mono_marshal_get_swift_physical_lowering`.
- changing returned lowering from `MonoTypeEnum` to `MonoType*` as this is the most common representation of types in the later Mono code flow
- fixing and refactoring algorithm for merging opaque intervals (previous version wasn't respecting correctly block boundaries)
- use correct field offset and value type size (taking into account `MONO_ABI_SIZEOF (MonoObject)`)
- fix `remaining_interval_size` for final lowering loop

## Verification
Currently, Mono CI only has coverage for JIT/interpreter but fullAOT verification was done locally. This PR passes an extended `SwiftAbiStressTest` of 5000 signature locally both on x64 and arm64 with JIT/interpreter/fullAOT backends.

## Alternatives
During preparation of this PR, several alternatives were explore and decided not to further proceed.
1. Implementing the lowering at the lower codegen level (`mini-<arch>.c` files). This option was discarded because Mini currently doesn't provide representation for value types that can be passed via registers and stack simultaneously. While handling it manually for this PR would be possibly, the final code was long and error prone. Additionally, the code would have to be duplicated for each architecture.
2. Implementing the lowering at IL-gen level (`marshal-lightweight.c`). This option was successfully implemented and provided a shared implementation for both Mini and interpreter. However, we decided to rather implement it at the Mono IR level. This has two benefits, firstly we get better control over the resulting assembly as we are closer to the Mono codegen. Secondly, we are able to keep interpreter code more independent for potential deployment of interpreter outside Mono.

---

FYI: @jakobbotsch  @amanasifkhalid @jkurdek @vitek-karas

matouskozak committed:
refactor + bug fixes for swift struct marshalling

matouskozak committed:
mini struct lowering

matouskozak committed:
interpreter swift struct lowering

matouskozak committed:
enable SwiftAbiStress on Mono

matouskozak committed:
explicit handling of byref swift structs

matouskozak committed:
move SwiftSelf and SwiftError checks
outside Swift struct lowering algorithm

matouskozak commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -3711,7 +3711,7 @@ mono_marshal_get_native_wrapper (MonoMethod *method, gboolean check_exceptions,
 						swift_error_args++;
 					} else if (param_klass == swift_self) {
 						swift_self_args++;
-					} else if (!m_class_is_blittable (param_klass) || m_class_is_simd_type (param_klass)) {
```
During AOT, class might not be properly initialize thus causing this check to fail even for blittable types.

lambdageek replied:
Can you call `mono_class_init_internal (mono_class_from_mono_type (method->signature->params[i]))` before checking if the type is blittable?

matouskozak replied:
The issue with AOT was that the wrappers don't get llvm compiled but are JIT/interpreter compiled during runtime and the blittable field was not correctly set at the time of this check. I believe calling class init wouldn't solve it, we would have to use `mono_class_setup_fields` (I *think*, haven't tried it). Do you think `mono_class_setup_fields (mono_class_from_mono_type (method->signature->params[i]))` would be better than just using `type_is_blittable (method->signature->params [i])`?

(I put the comment just to explain why I did the change, the code change in this PR makes it work with all codegens).

lambdageek replied:
yea try calling setup_fields before the blittable check

lambdageek replied:
I misunderstood the diff.  and I see `type_is_blittable` already calls `mono_class_init_sizes` which sets up fields when necessary.

lambdageek commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -3404,6 +3404,93 @@ interp_transform_call (TransformData *td, MonoMethod *method, MonoMethod *target
+		StackInfo *sp_old_params = (StackInfo*) mono_mempool_alloc (td->mempool, sizeof (StackInfo) * csignature->param_count);
+		memcpy (sp_old_params, td->sp, sizeof (StackInfo) * csignature->param_count);
+
+		GArray *new_params = g_array_new (FALSE, FALSE, sizeof (MonoType*));
```
Can you use `g_array_sized_new` with an initial size equal to the old param count (plus 4?).  that would avoid resizing most of the time.

matouskozak replied:
Yes, I changed it to use the old param count since we are unsure ahead-of-time if any lowering will occur.

lambdageek commented on src/mono/mono/mini/method-to-ir.c:
```diff
@@ -7514,6 +7514,89 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
+					}
+				}
+
+				// Create a new dummy signature with the lowered arguments
+				MonoMethodSignature *swiftcall_signature = mono_metadata_signature_alloc (m_class_get_image (method->klass), new_param_count);
+				for (uint32_t idx_param = 0; idx_param < new_param_count; ++idx_param) {
+					swiftcall_signature->params [idx_param] = g_array_index (new_params, MonoType *, idx_param);
+				}
+				swiftcall_signature->ret = fsig->ret;
+				swiftcall_signature->pinvoke = fsig->pinvoke;
+				swiftcall_signature->ext_callconv = fsig->ext_callconv;
+				
+				fsig = swiftcall_signature;	
```
1. Can we make a helper function and share it between the JIT and interp?
2. I'm not sure `m_class_get_image (method->klass)` is where we want to allocate the signature.  If the method belongs to a generic instance, we should use the mem manager of the generic instance (basically a mem manager that depends on the assemblies of the generic definition and the generic type params).

matouskozak replied:
Good point, I created a new helper function `mono_metadata_signature_dup_new_params` inside`metadata.c`.

lambdageek reviewed (commented):
mostly lgtm.  I'm not 100% sure about using memcpy on `StackInfo` in the interpreter.

I think (for the JIT especially) we need to be careful about where we allocate the new signature.

It would be nice if we could avoid `g_array_new` - can we have a faster path for, let's say, upto 16 arguments that uses a temporary stack allocation for the working array?

I think it might make sense to move the code for doing the lowering into a helper function in transform.c (and in method-to-ir.c) rather than doing everything inline in the giant switch statements.  That might keep the code a bit cleaner and easier to skip when someone is reading and they're not interested in Swift

kotlarmilos commented on src/mono/mono/mini/method-to-ir.c:
```diff
@@ -7514,6 +7514,89 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
+				n = fsig->param_count;
+				sp -= n;
+				// Save the old arguments				
+				MonoInst **old_args = g_newa (MonoInst*, n);
```
Deallocate `old_args` at the end.

matouskozak replied:
I changed it to using mempool allocation (`mono_mempool_alloc`) that should get deallocated automatically.

kotlarmilos commented on src/mono/mono/mini/method-to-ir.c:
```diff
@@ -7514,6 +7514,89 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
+
+				// Create a new dummy signature with the lowered arguments
+				MonoMethodSignature *swiftcall_signature = mono_metadata_signature_alloc (m_class_get_image (method->klass), new_param_count);
+				for (uint32_t idx_param = 0; idx_param < new_param_count; ++idx_param) {
```
Can the new signature be created in the "lowering" loop above?

matouskozak replied:
We would have to collect the lowering information before otherwise the final number of parameters is not known until all params are lowered. So we would still be doing essentially two loops.

kotlarmilos commented on src/mono/mono/mini/method-to-ir.c:
```diff
@@ -7514,6 +7514,89 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
+					old_args [idx_param] = sp [idx_param];
+				}
+
+				GArray *new_params = g_array_new (FALSE, FALSE, sizeof (MonoType*));
```
Deallocate if it is used.

kotlarmilos commented on src/mono/mono/mini/method-to-ir.c:
```diff
@@ -7514,6 +7514,89 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
+							MonoInst *struct_base_address =  mono_compile_create_var (cfg, mono_get_int_type (), OP_LOCAL);
+							CHECK_ARG (idx_param);
+							NEW_ARGLOADA (cfg, struct_base_address, idx_param);
+							MONO_ADD_INS (cfg->cbb, struct_base_address);
```
Will this be optimized in the final codegen?

matouskozak replied:
How optimized would you expect the final codegen to be? The resulting assembly looks somewhat like this:
```
...
000000000000007c        ldr     d0, [x29, #0x10] // Load first lowered element
0000000000000080        ldr     x0, [x29, #0x18] // Load second lowered element
0000000000000084        blr     x1 // Call to native Swift func
...
```
where `x29` contains the base address of the struct.

kotlarmilos replied:
As a new local variable is introduced here, and I assume that `x_arg`, where `_arg` is the arg index, is moved to `x29`. I wonder if the final codegen contains the `mov` ins or utilizes the `x_arg` directly.

matouskozak replied:
Based on the codegen, I think that it just uses an offset from the beginning of stack pointer (`x29`). It doesn't store the new local variable in registers.

kotlarmilos reviewed (commented):
Thanks for making changes verbose.

matouskozak committed:
Adding a helper function for new signature
The helper function duplicates signature metadata and adds new parameters to the signature.

+ pre-allocating new_params to the old params size

matouskozak committed:
extract swift struct lowering to separate function

Only for interpreter at the moment because mono_method_to_ir is more complicated due to NEW_ARGLOADA macros

matouskozak committed:
fix build warnings in signature alloc helper func

matouskozak committed:
deallocate temporary arrays

matouskozak committed:
fix cast to guint16 in transform.c

matouskozak commented:
Thank you for the feedback. I made several changes to improve the PR.

> mostly lgtm. I'm not 100% sure about using memcpy on `StackInfo` in the interpreter.

Done, removed the use of `memcpy` on `StackInfo`.

> I think (for the JIT especially) we need to be careful about where we allocate the new signature.

Done, changed the allocation of new signature to use `cfg->mempool` and `td->mempool` for mini and interpreter, respectively.

> It would be nice if we could avoid `g_array_new` - can we have a faster path for, let's say, upto 16 arguments that uses a temporary stack allocation for the working array?

We discussed that offline and agreed to stick with current dynamic allocation because we don't know the final number of parameters up front. In the future, we might change the process and collect the struct lowering information a little bit earlier (e.g. during marshalling https://github.com/dotnet/runtime/blob/19467dccf4e6786296eecd8007f90e6dafe01818/src/mono/mono/metadata/marshal.c#L3698-L3730) and save it in a temporary "swift-interop" mempool. Consequently, we would have more information for the lower layers and that could save up some allocations and speed up the code.

> I think it might make sense to move the code for doing the lowering into a helper function in transform.c (and in method-to-ir.c) rather than doing everything inline in the giant switch statements. That might keep the code a bit cleaner and easier to skip when someone is reading and they're not interested in Swift

Done for interpreter by creating `interp_emit_swiftcall_struct_lowering` helper function. I tried it for mini in `method-to-ir.c` as well but run into issues with increasing the shared stack pointer (`sp`) over the original amount of arguments.

jkoritzinsky reviewed (approved):
Changes to the lowering algorithm LGTM

kotlarmilos reviewed (approved):
LGTM!

BrzVlad commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -3320,6 +3320,88 @@ interp_try_devirt (MonoClass *this_klass, MonoMethod *target_method)
+{
+	MonoMethodSignature *new_csignature;
+	g_assert (!csignature->hasthis); // P/Invoke calls shouldn't contain 'this'
+
```
The argument moving around that happens here will break the fast offset allocator (where we allocate storage for vars immediately as we push it on the `td->sp` stack). This appears however to be fine because we disable tiering for wrappers and I would imagine a swift call is only happening from a m2n wrapper. However I would assert here that `td->optimized` with the comment that argument reordering here doesn't handle on the fly offset allocation and requires the full var offset allocator pass that is only ran for optimized code.

matouskozak replied:
Yes, we only need to perform the struct lowering for m2n (and later eventually n2m) wrappers. I added the assert with the comment you suggested.

matouskozak committed:
fix style in marshaling of swift structs

matouskozak committed:
assert td->optimized for interp struct lowering

matouskozak commented:
/azp run runtime-extra-platforms
